### PR TITLE
avoid information loss on .GoError()

### LINKED ISSUE #2628: Test failure in CI build 7857
The following test appears to have failed:

[#7857](https://circleci.com/gh/cockroachdb/cockroach/7857):

```
c962606727190855.13  127.0.0.1:0  22:56:46.727260  380.589µs  sending Batch            kv/txn_coord_sender.go:327
c962606727190855.13  127.0.0.1:0  22:56:46.727280  0          ·*proto.GetRequest       kv/txn_coord_sender.go:333
c962606727190855.13  127.0.0.1:0  22:56:46.727340  294.185µs  ·meta descriptor lookup  kv/dist_sender.go:510

I0922 22:56:46.730354 761 client/db.go:474  failed Batch: the descriptor for the first range is not available via gossip
--- FAIL: TestNodeStatusResponse (0.07s)
	status_test.go:270: the descriptor for the first range is not available via gossip
=== RUN   TestStoreStatusResponse
I0922 22:56:46.730779 761 base/context.go:155  setting up TLS from certificates directory: test_certs
I0922 22:56:46.735274 761 base/context.go:117  setting up TLS from certificates directory: test_certs
I0922 22:56:46.743322 761 rpc/clock_offset.go:155  monitoring cluster offset
I0922 22:56:46.743666 761 multiraft/multiraft.go:442  node 100000001 starting
I0922 22:56:46.744105 761 raft/raft.go:406  group 1 100000001 became follower at term 5
I0922 22:56:46.744269 761 raft/raft.go:219  group 1 newRaft 100000001 [peers: [100000001], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0922 22:56:46.744640 761 raft/raft.go:485  group 1 100000001 is starting a new election at term 5
I0922 22:56:46.744724 761 raft/raft.go:419  group 1 100000001 became candidate at term 6
--
2      ts/db.go:99
I0922 22:56:47.416030 761 stopper.go:237  draining; tasks left:
1      ts/db.go:99
I0922 22:56:47.438348 761 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestMetricsRecording (0.32s)
FAIL
FAIL	github.com/cockroachdb/cockroach/server	11.109s
=== RUN   TestNodeStatusMonitor
--- PASS: TestNodeStatusMonitor (0.01s)
=== RUN   TestNodeStatusRecorder
--- PASS: TestNodeStatusRecorder (0.01s)
=== RUN   TestRuntimeStatRecorder
I0922 22:56:52.692566 797 server/status/runtime.go:118  runtime stats: 10 goroutines, 5.43MiB active, 50000000.00cgo/sec, 340630.00/113540.00 %(u/s)time, 1704.57 %gc (1x)
--- PASS: TestRuntimeStatRecorder (0.00s)
=== RUN   TestNodeEventFeed
--- PASS: TestNodeEventFeed (0.00s)
=== RUN   TestServerNodeEventFeed

```
Please assign, take a look and update the issue accordingly.

### PR DESCRIPTION
**Pull request by tbg**
fixes #2628.

### CODE REVIEW

File: proto/errors.go [hunk 1]
```diff
@@ -67,7 +72,42 @@ func NewError(err error) *Error {
...
+func (e *internalError) Transaction() *Transaction {
+	return nil
+}
+
+// ErrorIndex implements IndexedError.
+func (e *internalError) ErrorIndex() (int32, bool) {
+	if e.Index == nil {
+		return 0, false
+	}
+	return (*Error)(e).Index.Index, true
```
Comment (tamird) [inline proto/errors.go]: why bother with the cast?

File: storage/store_test.go [hunk 2]
```diff
@@ -1550,3 +1551,19 @@ func TestMaybeRemove(t *testing.T) {
...
+
+func TestErrWithIndex(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	ind := int32(99)
+	err := error(&errWithIndex{err: errors.New("foo"), index: ind})
+	i1, ok1 := err.(proto.IndexedError).ErrorIndex()
+	pErr := &proto.Error{}
+	pErr.SetResponseGoError(err)
+	if pErr.GetIndex().GetIndex() != ind {
+		log.Warningf("proto.Error lost index information")
```
Comment (tamird) [inline storage/store_test.go]: whats this about?

Comment (tschottdorf) [inline storage/store_test.go]: this just makes sure that `SetResponseGoError()` will set the `Index` in the `*proto.Error` message. If `err` implemented `ErrorDetail` (doesn't now, but could in the future), the next test using `GoError()` below would get the information from the original error, but it should do the same in either case.

File: proto/errors.go (hunk 1 above)
Comment (tschottdorf) [inline proto/errors.go]: I think I used `GetIndex()` before. Changed.

**Comment by tschottdorf** [comment]
PTAL. I actually cleaned up some more stuff in the vicinity (see third commit message), which brought some nice simplifications.

**Comment by tamird** [comment]
That these errors carry indexes into the batch in which they are contained is starting to feel pretty dirty to me. What's the long term plan here?

File: storage/store_test.go [hunk 3]
```diff
@@ -465,8 +465,6 @@ func TestStoreExecuteNoop(t *testing.T) {
 	}
 }
 
-// TestStoreErrorWithIndex verifies that an *errWithIndex returned from the
-// replica is correctly translated into a *proto.Error with a corresponding
 // Index. See the companion test TestBatchWithError.
```
Comment (tamird) [inline storage/store_test.go]: this comment is broken now

File: proto/errors.go [hunk 4]
```diff
@@ -113,13 +175,22 @@ func (e *Error) SetResponseGoError(err error) {
 	if r, ok := err.(retry.Retryable); ok {
 		e.Retryable = r.CanRetry()
 	}
+	var isTxnError bool
 	if r, ok := err.(TransactionRestartError); ok {
```
Comment (tamird) [inline proto/errors.go]: s/ok/isTxnError/

**Comment by tamird** [comment]
LGTM

File: proto/errors.proto [hunk 5]
```diff
@@ -192,6 +193,11 @@ enum TransactionRestart {
   IMMEDIATE = 2;
 }
 
+// ErrPosition describes the position of an error in a Batch.
+message ErrPosition {
```
Comment (petermattis) [inline proto/errors.proto]: Is there a value to this message versus just embedding `optional int32 index` in the errors which require it?

**Comment by petermattis** [comment]
LGTM

File: proto/errors.proto (hunk 5 above)
Comment (tamird) [inline proto/errors.proto]: yes, in proto3 primitive types are not nullable. there was a discussion on this in a previous PR

**Comment by bdarnell** [comment]
LGTM. This change is definitely an improvement, but embedding the position doesn't sit well with me either. Is there something else we could do to provide the same functionality? For example, if ConditionalPut returned the key as well as the value, would that be enough for the SQL layer to construct the right error message for uniqueness violations?

**Comment by tamird** [comment]
@bdarnell yes, that would be enough for the SQL layer. However, I think in general we will still want an error -> operation mapping for batches.

**Comment by tschottdorf** [comment]
The only location in which they're really "on" the batch is when they're on
the wire, but everywhere else they're separate. In the end, I would like to
completely remove the error field from `*proto.BatchResponse` and just send
a `proto.BatchResponseOrError` over the wire, so that you never see both a
response and an error.

On Wed, Sep 23, 2015 at 10:52 AM, Tamir Duberstein <notifications@github.com

File: storage/store_test.go (hunk 3 above)
Comment (tschottdorf) [inline storage/store_test.go]: Removed the test. `TestBatchErrorWithIndex` tests the same thing closer to
the source.

On Wed, Sep 23, 2015 at 10:53 AM, Tamir Duberstein <notifications@github.com

File: proto/errors.go (hunk 4 above)
Comment (tschottdorf) [inline proto/errors.go]: I need it defined in the outside scope.

File: proto/errors.proto (hunk 5 above)
Comment (tschottdorf) [inline proto/errors.proto]: I've moved up the comment about that from `message Error`.

**Comment by tschottdorf** [comment]
> For example, if ConditionalPut returned the key as well as the value

Yes, I thought about that too but it seemed fairly inflexible. With hindsight, while I think this change makes indexed errors workable, I agree it's not the cream of the crop yet and it might've been better to start small and just add a `Key` to `proto.ConditionalPutError`. Well, damage is done now. I'm sure there's going to be another iteration.

File: proto/errors.go (hunk 4 above)
Comment (tamird) [inline proto/errors.go]: yeah i understand, but you can do that by just:
```go
r, isTxnError := err.(TransactionRestartError)
if isTxnError {
  e.TransactionRestart = r.CanRestartTransaction()
}
```

Comment (tschottdorf) [inline proto/errors.go]: going to leave as is to keep the symmetry with the other block.

**Pull request merged**